{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ac97d5be-6379-434c-8158-e9c94e780bbd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "import plotly.io as pio\n",
    "from typing import Dict, List, Tuple\n",
    "import torch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b74d84dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__bool__',\n",
       " '__class__',\n",
       " '__delattr__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__le__',\n",
       " '__lt__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__subclasshook__']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(pio.kaleido.scope)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "52b686c5",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'default_format'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[49], line 6\u001b[0m\n\u001b[1;32m      4\u001b[0m pio\u001b[39m.\u001b[39mrenderers\u001b[39m.\u001b[39mdefault\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mjupyterlab\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m      5\u001b[0m \u001b[39m# pio.renderers.default = 'iframe_connected'\u001b[39;00m\n\u001b[0;32m----> 6\u001b[0m pio\u001b[39m.\u001b[39;49mkaleido\u001b[39m.\u001b[39;49mscope\u001b[39m.\u001b[39;49mdefault_format \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39msvg\u001b[39m\u001b[39m\"\u001b[39m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'default_format'"
     ]
    }
   ],
   "source": [
    "# https://plotly.com/python/static-image-export/\n",
    "import plotly.io as pio\n",
    "# pio.renderers.default='notebook'\n",
    "pio.renderers.default='jupyterlab'\n",
    "# pio.renderers.default = 'iframe_connected'\n",
    "pio.kaleido.scope.default_format = \"svg\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "12e4b70e-5c33-4261-8a68-795ff4d30890",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "app = \"/app\"\n",
    "if app not in sys.path:\n",
    "    sys.path.append(app)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0bd3ab2b-733f-49c8-b88e-5e139865cd55",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from src.comparator import Comparator\n",
    "from src.embed import OpenAIEmbedder, SBERTEmbedder, BarlowEmbedder\n",
    "from src.preprint import Preprint\n",
    "from src.review_process import ReviewProcess\n",
    "from src.utils import split_paragraphs, split_sentences\n",
    "from src.config import config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "dddf7e02-58a7-43c6-8398-f287d075f890",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Config(embedding_model={'openai': 'text-embedding-ada-002', 'sbert': 'all-mpnet-base-v2', 'barlow': '/app/pretrained/twin-lm-checkpoint-32000'}, sections='introduction+results+discussion+methods')"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "017fc49b-18f4-46d8-af8f-65f7c50d774a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "doi = \"10.1101/2021.05.12.443743\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4469b9d5-8aee-44f9-b15d-8d60a64a32c5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "preprint = Preprint(doi=doi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "aaa926dd-1754-454c-b4d1-6968af1a4147",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "review_process = ReviewProcess(doi=doi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f326cd9f-2f29-4ffa-9a63-4978e7dbc21f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "preprint_chunks = preprint.get_chunks(split_paragraphs)\n",
    "# preprint_chunks = preprint.get_chunks(split_sentences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f0e819ed-3351-4738-a8b3-9516f747344c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "review_1_chunks = review_process.reviews[0].get_chunks(split_paragraphs)\n",
    "# review_chunks = review_process.reviews[0].get_chunks(split_sentences)\n",
    "review_2_chunks = review_process.reviews[1].get_chunks(split_paragraphs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f6b6ee30-79e1-4164-bd26-02ce417ad375",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "preprint chunks: 52; review chunks: 17\n"
     ]
    }
   ],
   "source": [
    "print(f\"preprint chunks: {len(preprint_chunks)}; review chunks: {len(review_1_chunks)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "33cd43cc-d898-498d-8734-c9183eda20ce",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chosen=text-embedding-ada-002\n"
     ]
    }
   ],
   "source": [
    "# embedder = SBERTEmbedder()\n",
    "embedder = OpenAIEmbedder()\n",
    "# embedder = BarlowEmbedder(model=\"/app/pretrained/twin-lm-checkpoint-32000\", mode='paragraph')\n",
    "\n",
    "print(f\"chosen={embedder.model}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "53db513a-c9b9-4f15-a6b6-5996278766c2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "comp = Comparator(embedder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "000b698f-d7a8-43d2-b568-b897f53d2529",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "similarity_matrix = comp.compare_dot(review_1_chunks, preprint_chunks) #preprint_chunks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "de228f71-ef8d-4689-92b6-1b6b9ac13c69",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "torch.Size([17, 52])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "similarity_matrix.size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1bf8206d-e68e-4fdd-81eb-93694a700397",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>                            <div id=\"1b200369-dedd-401d-9410-e1359ddca891\" class=\"plotly-graph-div\" style=\"height:500px; width:1000px;\"></div>            <script type=\"text/javascript\">                require([\"plotly\"], function(Plotly) {                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById(\"1b200369-dedd-401d-9410-e1359ddca891\")) {                    Plotly.newPlot(                        \"1b200369-dedd-401d-9410-e1359ddca891\",                        [{\"coloraxis\":\"coloraxis\",\"name\":\"0\",\"z\":[[0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929],[0.8619670271873474,0.8542410135269165,0.9074594378471375,0.8828065395355225,0.8929170370101929,0.8755152821540833,0.9099065065383911,0.9204180240631104,0.902197003364563,0.94364333152771,0.9409996271133423,0.9304534792900085,0.8895666599273682,0.9122796654701233,0.906971275806427,0.8600445985794067,0.8831936717033386,0.8843904137611389,0.8880491852760315,0.8553113341331482,0.8801369667053223,0.8534400463104248,0.856918215751648,0.8497638702392578,0.8731714487075806,0.8914791345596313,0.8980767726898193,0.8338832259178162,0.9075753688812256,0.9366087913513184,0.914405345916748,0.8681038618087769,0.8640804290771484,0.8516395092010498,0.8703033924102783,0.800000011920929,0.862791895866394,0.834123969078064,0.8429787158966064,0.8268735408782959,0.8076901435852051,0.8866308927536011,0.8428429961204529,0.8571990728378296,0.9042980670928955,0.8665180802345276,0.8482666015625,0.8367217779159546,0.8300386071205139,0.8316177129745483,0.800000011920929,0.800000011920929],[0.8602150082588196,0.8442690372467041,0.9287459850311279,0.9169116020202637,0.9196897745132446,0.8911712765693665,0.9184557199478149,0.9146350026130676,0.8656124472618103,0.9277056455612183,0.934516429901123,0.943965494632721,0.8715378642082214,0.9211054444313049,0.9033827781677246,0.8284586668014526,0.8763197064399719,0.8998447060585022,0.8945504426956177,0.8347938060760498,0.8980935215950012,0.8443723320960999,0.8312580585479736,0.8469711542129517,0.8577507734298706,0.8944941163063049,0.8911064267158508,0.8008139133453369,0.9077455997467041,0.937423825263977,0.9411038756370544,0.8849619030952454,0.8824658393859863,0.8532018661499023,0.8991022706031799,0.800000011920929,0.815933346748352,0.8002651929855347,0.800000011920929,0.800000011920929,0.800000011920929,0.8558449745178223,0.8179656863212585,0.8295340538024902,0.8822315335273743,0.838111162185669,0.8058321475982666,0.8146759867668152,0.8068366646766663,0.800000011920929,0.800000011920929,0.800000011920929],[0.8749874830245972,0.8674519062042236,0.9135681390762329,0.9264273643493652,0.9190073013305664,0.9248663783073425,0.9335552453994751,0.9087932109832764,0.8610181212425232,0.8989323377609253,0.9041523337364197,0.9018900990486145,0.8363651037216187,0.9043350219726562,0.8948814868927002,0.8303921818733215,0.8650718927383423,0.9133602380752563,0.9121180772781372,0.8618966341018677,0.9148073792457581,0.9060739278793335,0.8575223684310913,0.902436375617981,0.8606351613998413,0.9159876704216003,0.9228329658508301,0.8217229843139648,0.9277167320251465,0.9059386849403381,0.9213635921478271,0.9117226600646973,0.9267445802688599,0.897824764251709,0.9236484169960022,0.800000011920929,0.8208274841308594,0.8260194063186646,0.8088383674621582,0.8049430847167969,0.800000011920929,0.8506584167480469,0.8071732521057129,0.8516600131988525,0.8677008152008057,0.8584878444671631,0.8098684549331665,0.8281297087669373,0.8160227537155151,0.800000011920929,0.800000011920929,0.800000011920929],[0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929],[0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929],[0.8422486782073975,0.829856812953949,0.851408064365387,0.8364808559417725,0.8412020206451416,0.8419094085693359,0.863634467124939,0.8539972305297852,0.873346209526062,0.8991067409515381,0.8984493017196655,0.8877971172332764,0.8666845560073853,0.87678062915802,0.9323590397834778,0.8986988067626953,0.920601487159729,0.8869592547416687,0.8474152088165283,0.8369251489639282,0.8535146713256836,0.8232470750808716,0.8287347555160522,0.8156366348266602,0.856040358543396,0.8478695154190063,0.8708989024162292,0.829221248626709,0.8793776631355286,0.8977971076965332,0.8628712892532349,0.8691906332969666,0.819979727268219,0.8383586406707764,0.8508467674255371,0.800000011920929,0.8124631643295288,0.800000011920929,0.8067144155502319,0.8027832508087158,0.800745964050293,0.8566524982452393,0.833939254283905,0.8745341897010803,0.8988275527954102,0.8880096673965454,0.8179963827133179,0.8260437250137329,0.8351473808288574,0.8156710863113403,0.800000011920929,0.800000011920929],[0.8833935856819153,0.8671610355377197,0.882423460483551,0.912966787815094,0.8958109617233276,0.8862487077713013,0.8859072923660278,0.855617880821228,0.8330589532852173,0.8525633215904236,0.852949857711792,0.8521218299865723,0.8051785826683044,0.8708515167236328,0.8755241632461548,0.8228241205215454,0.8472759127616882,0.8824970722198486,0.8840222358703613,0.836263120174408,0.8733229041099548,0.858551561832428,0.8133865594863892,0.8524538278579712,0.8174123764038086,0.8525002002716064,0.8895779848098755,0.800000011920929,0.896350622177124,0.8487900495529175,0.8648804426193237,0.8925554156303406,0.8778770565986633,0.8897190093994141,0.9047354459762573,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.8217261433601379,0.800000011920929,0.8295959234237671,0.8280285000801086,0.8499393463134766,0.800000011920929,0.8142189979553223,0.8084062337875366,0.800000011920929,0.800000011920929,0.800000011920929],[0.8436846733093262,0.8372731804847717,0.858832061290741,0.8471567034721375,0.849088191986084,0.855377733707428,0.854650616645813,0.8355731964111328,0.8312339782714844,0.8714820742607117,0.8829609155654907,0.8676669597625732,0.8575999736785889,0.8729652762413025,0.90498948097229,0.8711190223693848,0.8823884725570679,0.8890992403030396,0.8381937742233276,0.815453052520752,0.8587051630020142,0.8266471028327942,0.8060486316680908,0.8272877931594849,0.8478661775588989,0.8379454612731934,0.869498610496521,0.8134350776672363,0.8802177906036377,0.8756625652313232,0.8684976100921631,0.8634800910949707,0.8315834403038025,0.8347424268722534,0.8575953245162964,0.800000011920929,0.8083130717277527,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.8395206332206726,0.8412488102912903,0.8640198111534119,0.8893771171569824,0.8715097904205322,0.8233615159988403,0.8210960030555725,0.8345833420753479,0.800000011920929,0.800000011920929,0.800000011920929],[0.8302720189094543,0.849384069442749,0.8402149677276611,0.8371017575263977,0.8420602679252625,0.8736518025398254,0.8649677038192749,0.8823026418685913,0.8873211145401001,0.8721786737442017,0.8767713308334351,0.8409017324447632,0.8339954614639282,0.8540740013122559,0.8694750070571899,0.8491255640983582,0.845686674118042,0.8519273996353149,0.877895712852478,0.8718182444572449,0.8395114541053772,0.8855915069580078,0.878983199596405,0.8863620162010193,0.8555556535720825,0.8812048435211182,0.9145548939704895,0.8914048671722412,0.9268089532852173,0.8578771352767944,0.8457668423652649,0.8491827249526978,0.8600821495056152,0.857624888420105,0.8749409914016724,0.800000011920929,0.8433008193969727,0.8436416983604431,0.8500497341156006,0.8756442070007324,0.8432638645172119,0.8487623929977417,0.8135049939155579,0.8636033535003662,0.8545372486114502,0.8521409034729004,0.8249561190605164,0.825729250907898,0.8233463764190674,0.8147048354148865,0.8096271753311157,0.800000011920929],[0.8340649604797363,0.8402810096740723,0.8549250364303589,0.873984694480896,0.8748828172683716,0.9075405597686768,0.8841677308082581,0.8914945721626282,0.865207314491272,0.8702970147132874,0.8808353543281555,0.8619050979614258,0.8274959325790405,0.867304801940918,0.8517025709152222,0.8099265694618225,0.8288724422454834,0.8598758578300476,0.8731153607368469,0.8450647592544556,0.8934193849563599,0.8818546533584595,0.8412263989448547,0.877622127532959,0.830525279045105,0.8824803829193115,0.9165655970573425,0.867982029914856,0.9162889719009399,0.8629549145698547,0.8583749532699585,0.8637888431549072,0.8969826698303223,0.8822276592254639,0.9024913311004639,0.800000011920929,0.8148393630981445,0.8196027278900146,0.8171911239624023,0.8537650108337402,0.800000011920929,0.8424740433692932,0.800000011920929,0.8157336711883545,0.8473119735717773,0.8367896676063538,0.8031171560287476,0.8069138526916504,0.8007115721702576,0.800000011920929,0.8068331480026245,0.800000011920929],[0.800000011920929,0.800000011920929,0.8148438930511475,0.800000011920929,0.83831787109375,0.8007198572158813,0.8057923913002014,0.8194238543510437,0.8280763030052185,0.8332267999649048,0.8644437789916992,0.8537640571594238,0.8691103458404541,0.8763024210929871,0.8329247832298279,0.8478556871414185,0.8505492806434631,0.8264484405517578,0.8222453594207764,0.8171953558921814,0.8000114560127258,0.800000011920929,0.802968442440033,0.800000011920929,0.8249003887176514,0.8142646551132202,0.8198300004005432,0.8031307458877563,0.8515526056289673,0.8216937780380249,0.8407232165336609,0.8174439668655396,0.800000011920929,0.800000011920929,0.8075287342071533,0.800000011920929,0.800000011920929,0.800000011920929,0.8000045418739319,0.800000011920929,0.800000011920929,0.8281093239784241,0.800000011920929,0.800000011920929,0.8316556215286255,0.8265036344528198,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929],[0.8064248561859131,0.800000011920929,0.8032121658325195,0.800000011920929,0.8005939722061157,0.800000011920929,0.813422679901123,0.800000011920929,0.8302239775657654,0.824164092540741,0.8290717601776123,0.800000011920929,0.8354401588439941,0.8236072063446045,0.8207302689552307,0.8555346131324768,0.8367210626602173,0.8131438493728638,0.8131538033485413,0.8191061615943909,0.800000011920929,0.800000011920929,0.8018274307250977,0.800000011920929,0.830675482749939,0.800000011920929,0.8170245885848999,0.8059179782867432,0.8101663589477539,0.8059568405151367,0.800000011920929,0.8148280382156372,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.8140542507171631,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.8002946972846985,0.8339574337005615,0.8289024829864502,0.8258163928985596,0.8249580264091492,0.8342768549919128,0.8608875274658203,0.8716930150985718,0.8111668825149536,0.800000011920929,0.800000011920929],[0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.8226407766342163,0.8021317720413208,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929],[0.8592569828033447,0.8495529294013977,0.9403008818626404,0.9069609642028809,0.9175832271575928,0.8952208161354065,0.9042769074440002,0.8958277702331543,0.8549026250839233,0.8936072587966919,0.8913688659667969,0.8854171633720398,0.8367505073547363,0.8922591209411621,0.8614215850830078,0.8121543526649475,0.8315119743347168,0.8659572601318359,0.8869878053665161,0.8392486572265625,0.8641153573989868,0.8496625423431396,0.8351649045944214,0.8632587194442749,0.8464929461479187,0.878303050994873,0.8784065246582031,0.800000011920929,0.8797555565834045,0.882473886013031,0.914802074432373,0.8806136846542358,0.8899146318435669,0.866320788860321,0.878982424736023,0.800000011920929,0.8131281137466431,0.806296706199646,0.800000011920929,0.800000011920929,0.800000011920929,0.8345428705215454,0.800000011920929,0.8199004530906677,0.8539283871650696,0.8313534259796143,0.8232607841491699,0.8093840479850769,0.8064202666282654,0.800000011920929,0.800000011920929,0.800000011920929],[0.8626303672790527,0.8514048457145691,0.9250853061676025,0.9338843822479248,0.9402590990066528,0.9235514402389526,0.9549608826637268,0.9165002703666687,0.8625088930130005,0.9041341543197632,0.905697226524353,0.8996443152427673,0.8397918939590454,0.9032946825027466,0.8769989013671875,0.8245582580566406,0.84010910987854,0.8791489601135254,0.9112937450408936,0.8495206832885742,0.8952757716178894,0.8799114227294922,0.8512458801269531,0.8927820920944214,0.8624744415283203,0.902912974357605,0.9099286794662476,0.8036606311798096,0.9019752144813538,0.9157712459564209,0.9188557863235474,0.8868266940116882,0.9236798882484436,0.8851184844970703,0.9243611693382263,0.800000011920929,0.8188312649726868,0.8224321603775024,0.800000011920929,0.800000011920929,0.800000011920929,0.8249505758285522,0.800000011920929,0.8400494456291199,0.8616143465042114,0.8496227264404297,0.8274857401847839,0.816757082939148,0.8111237287521362,0.800000011920929,0.800000011920929,0.800000011920929],[0.8106663227081299,0.8084027171134949,0.8428531289100647,0.8198768496513367,0.8284487724304199,0.8426294922828674,0.8346330523490906,0.8196232318878174,0.8125841021537781,0.8258165121078491,0.8248331546783447,0.8152475953102112,0.800000011920929,0.8212416172027588,0.8039213418960571,0.800000011920929,0.800000011920929,0.8035281896591187,0.8197818994522095,0.800000011920929,0.8297618627548218,0.8025873899459839,0.800000011920929,0.8026295900344849,0.800000011920929,0.8154950141906738,0.8273818492889404,0.800000011920929,0.8134160041809082,0.8244867920875549,0.8254072070121765,0.8092862367630005,0.8290113806724548,0.811893105506897,0.8021944761276245,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.8030042052268982,0.8086873888969421,0.8004689812660217,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929,0.800000011920929]],\"type\":\"heatmap\",\"xaxis\":\"x\",\"yaxis\":\"y\",\"hovertemplate\":\"x: %{x}<br>y: %{y}<br>color: %{z}<extra></extra>\"}],                        {\"template\":{\"data\":{\"barpolar\":[{\"marker\":{\"line\":{\"color\":\"rgb(234,234,242)\",\"width\":0.5},\"pattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2}},\"type\":\"barpolar\"}],\"bar\":[{\"error_x\":{\"color\":\"rgb(36,36,36)\"},\"error_y\":{\"color\":\"rgb(36,36,36)\"},\"marker\":{\"line\":{\"color\":\"rgb(234,234,242)\",\"width\":0.5},\"pattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2}},\"type\":\"bar\"}],\"carpet\":[{\"aaxis\":{\"endlinecolor\":\"rgb(36,36,36)\",\"gridcolor\":\"white\",\"linecolor\":\"white\",\"minorgridcolor\":\"white\",\"startlinecolor\":\"rgb(36,36,36)\"},\"baxis\":{\"endlinecolor\":\"rgb(36,36,36)\",\"gridcolor\":\"white\",\"linecolor\":\"white\",\"minorgridcolor\":\"white\",\"startlinecolor\":\"rgb(36,36,36)\"},\"type\":\"carpet\"}],\"choropleth\":[{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2},\"type\":\"choropleth\"}],\"contourcarpet\":[{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2},\"type\":\"contourcarpet\"}],\"contour\":[{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2},\"colorscale\":[[0.0,\"rgb(2,4,25)\"],[0.06274509803921569,\"rgb(24,15,41)\"],[0.12549019607843137,\"rgb(47,23,57)\"],[0.18823529411764706,\"rgb(71,28,72)\"],[0.25098039215686274,\"rgb(97,30,82)\"],[0.3137254901960784,\"rgb(123,30,89)\"],[0.3764705882352941,\"rgb(150,27,91)\"],[0.4392156862745098,\"rgb(177,22,88)\"],[0.5019607843137255,\"rgb(203,26,79)\"],[0.5647058823529412,\"rgb(223,47,67)\"],[0.6274509803921569,\"rgb(236,76,61)\"],[0.6901960784313725,\"rgb(242,107,73)\"],[0.7529411764705882,\"rgb(244,135,95)\"],[0.8156862745098039,\"rgb(245,162,122)\"],[0.8784313725490196,\"rgb(246,188,153)\"],[0.9411764705882353,\"rgb(247,212,187)\"],[1.0,\"rgb(250,234,220)\"]],\"type\":\"contour\"}],\"heatmapgl\":[{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2},\"colorscale\":[[0.0,\"rgb(2,4,25)\"],[0.06274509803921569,\"rgb(24,15,41)\"],[0.12549019607843137,\"rgb(47,23,57)\"],[0.18823529411764706,\"rgb(71,28,72)\"],[0.25098039215686274,\"rgb(97,30,82)\"],[0.3137254901960784,\"rgb(123,30,89)\"],[0.3764705882352941,\"rgb(150,27,91)\"],[0.4392156862745098,\"rgb(177,22,88)\"],[0.5019607843137255,\"rgb(203,26,79)\"],[0.5647058823529412,\"rgb(223,47,67)\"],[0.6274509803921569,\"rgb(236,76,61)\"],[0.6901960784313725,\"rgb(242,107,73)\"],[0.7529411764705882,\"rgb(244,135,95)\"],[0.8156862745098039,\"rgb(245,162,122)\"],[0.8784313725490196,\"rgb(246,188,153)\"],[0.9411764705882353,\"rgb(247,212,187)\"],[1.0,\"rgb(250,234,220)\"]],\"type\":\"heatmapgl\"}],\"heatmap\":[{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2},\"colorscale\":[[0.0,\"rgb(2,4,25)\"],[0.06274509803921569,\"rgb(24,15,41)\"],[0.12549019607843137,\"rgb(47,23,57)\"],[0.18823529411764706,\"rgb(71,28,72)\"],[0.25098039215686274,\"rgb(97,30,82)\"],[0.3137254901960784,\"rgb(123,30,89)\"],[0.3764705882352941,\"rgb(150,27,91)\"],[0.4392156862745098,\"rgb(177,22,88)\"],[0.5019607843137255,\"rgb(203,26,79)\"],[0.5647058823529412,\"rgb(223,47,67)\"],[0.6274509803921569,\"rgb(236,76,61)\"],[0.6901960784313725,\"rgb(242,107,73)\"],[0.7529411764705882,\"rgb(244,135,95)\"],[0.8156862745098039,\"rgb(245,162,122)\"],[0.8784313725490196,\"rgb(246,188,153)\"],[0.9411764705882353,\"rgb(247,212,187)\"],[1.0,\"rgb(250,234,220)\"]],\"type\":\"heatmap\"}],\"histogram2dcontour\":[{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2},\"colorscale\":[[0.0,\"rgb(2,4,25)\"],[0.06274509803921569,\"rgb(24,15,41)\"],[0.12549019607843137,\"rgb(47,23,57)\"],[0.18823529411764706,\"rgb(71,28,72)\"],[0.25098039215686274,\"rgb(97,30,82)\"],[0.3137254901960784,\"rgb(123,30,89)\"],[0.3764705882352941,\"rgb(150,27,91)\"],[0.4392156862745098,\"rgb(177,22,88)\"],[0.5019607843137255,\"rgb(203,26,79)\"],[0.5647058823529412,\"rgb(223,47,67)\"],[0.6274509803921569,\"rgb(236,76,61)\"],[0.6901960784313725,\"rgb(242,107,73)\"],[0.7529411764705882,\"rgb(244,135,95)\"],[0.8156862745098039,\"rgb(245,162,122)\"],[0.8784313725490196,\"rgb(246,188,153)\"],[0.9411764705882353,\"rgb(247,212,187)\"],[1.0,\"rgb(250,234,220)\"]],\"type\":\"histogram2dcontour\"}],\"histogram2d\":[{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2},\"colorscale\":[[0.0,\"rgb(2,4,25)\"],[0.06274509803921569,\"rgb(24,15,41)\"],[0.12549019607843137,\"rgb(47,23,57)\"],[0.18823529411764706,\"rgb(71,28,72)\"],[0.25098039215686274,\"rgb(97,30,82)\"],[0.3137254901960784,\"rgb(123,30,89)\"],[0.3764705882352941,\"rgb(150,27,91)\"],[0.4392156862745098,\"rgb(177,22,88)\"],[0.5019607843137255,\"rgb(203,26,79)\"],[0.5647058823529412,\"rgb(223,47,67)\"],[0.6274509803921569,\"rgb(236,76,61)\"],[0.6901960784313725,\"rgb(242,107,73)\"],[0.7529411764705882,\"rgb(244,135,95)\"],[0.8156862745098039,\"rgb(245,162,122)\"],[0.8784313725490196,\"rgb(246,188,153)\"],[0.9411764705882353,\"rgb(247,212,187)\"],[1.0,\"rgb(250,234,220)\"]],\"type\":\"histogram2d\"}],\"histogram\":[{\"marker\":{\"pattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2}},\"type\":\"histogram\"}],\"mesh3d\":[{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2},\"type\":\"mesh3d\"}],\"parcoords\":[{\"line\":{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2}},\"type\":\"parcoords\"}],\"pie\":[{\"automargin\":true,\"type\":\"pie\"}],\"scatter3d\":[{\"line\":{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2}},\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2}},\"type\":\"scatter3d\"}],\"scattercarpet\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2}},\"type\":\"scattercarpet\"}],\"scattergeo\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2}},\"type\":\"scattergeo\"}],\"scattergl\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2}},\"type\":\"scattergl\"}],\"scattermapbox\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2}},\"type\":\"scattermapbox\"}],\"scatterpolargl\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2}},\"type\":\"scatterpolargl\"}],\"scatterpolar\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2}},\"type\":\"scatterpolar\"}],\"scatter\":[{\"fillpattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2},\"type\":\"scatter\"}],\"scatterternary\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2}},\"type\":\"scatterternary\"}],\"surface\":[{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2},\"colorscale\":[[0.0,\"rgb(2,4,25)\"],[0.06274509803921569,\"rgb(24,15,41)\"],[0.12549019607843137,\"rgb(47,23,57)\"],[0.18823529411764706,\"rgb(71,28,72)\"],[0.25098039215686274,\"rgb(97,30,82)\"],[0.3137254901960784,\"rgb(123,30,89)\"],[0.3764705882352941,\"rgb(150,27,91)\"],[0.4392156862745098,\"rgb(177,22,88)\"],[0.5019607843137255,\"rgb(203,26,79)\"],[0.5647058823529412,\"rgb(223,47,67)\"],[0.6274509803921569,\"rgb(236,76,61)\"],[0.6901960784313725,\"rgb(242,107,73)\"],[0.7529411764705882,\"rgb(244,135,95)\"],[0.8156862745098039,\"rgb(245,162,122)\"],[0.8784313725490196,\"rgb(246,188,153)\"],[0.9411764705882353,\"rgb(247,212,187)\"],[1.0,\"rgb(250,234,220)\"]],\"type\":\"surface\"}],\"table\":[{\"cells\":{\"fill\":{\"color\":\"rgb(231,231,240)\"},\"line\":{\"color\":\"white\"}},\"header\":{\"fill\":{\"color\":\"rgb(183,183,191)\"},\"line\":{\"color\":\"white\"}},\"type\":\"table\"}]},\"layout\":{\"annotationdefaults\":{\"arrowcolor\":\"rgb(67,103,167)\"},\"autotypenumbers\":\"strict\",\"coloraxis\":{\"colorbar\":{\"outlinewidth\":0,\"tickcolor\":\"rgb(36,36,36)\",\"ticklen\":8,\"ticks\":\"outside\",\"tickwidth\":2}},\"colorscale\":{\"sequential\":[[0.0,\"rgb(2,4,25)\"],[0.06274509803921569,\"rgb(24,15,41)\"],[0.12549019607843137,\"rgb(47,23,57)\"],[0.18823529411764706,\"rgb(71,28,72)\"],[0.25098039215686274,\"rgb(97,30,82)\"],[0.3137254901960784,\"rgb(123,30,89)\"],[0.3764705882352941,\"rgb(150,27,91)\"],[0.4392156862745098,\"rgb(177,22,88)\"],[0.5019607843137255,\"rgb(203,26,79)\"],[0.5647058823529412,\"rgb(223,47,67)\"],[0.6274509803921569,\"rgb(236,76,61)\"],[0.6901960784313725,\"rgb(242,107,73)\"],[0.7529411764705882,\"rgb(244,135,95)\"],[0.8156862745098039,\"rgb(245,162,122)\"],[0.8784313725490196,\"rgb(246,188,153)\"],[0.9411764705882353,\"rgb(247,212,187)\"],[1.0,\"rgb(250,234,220)\"]],\"sequentialminus\":[[0.0,\"rgb(2,4,25)\"],[0.06274509803921569,\"rgb(24,15,41)\"],[0.12549019607843137,\"rgb(47,23,57)\"],[0.18823529411764706,\"rgb(71,28,72)\"],[0.25098039215686274,\"rgb(97,30,82)\"],[0.3137254901960784,\"rgb(123,30,89)\"],[0.3764705882352941,\"rgb(150,27,91)\"],[0.4392156862745098,\"rgb(177,22,88)\"],[0.5019607843137255,\"rgb(203,26,79)\"],[0.5647058823529412,\"rgb(223,47,67)\"],[0.6274509803921569,\"rgb(236,76,61)\"],[0.6901960784313725,\"rgb(242,107,73)\"],[0.7529411764705882,\"rgb(244,135,95)\"],[0.8156862745098039,\"rgb(245,162,122)\"],[0.8784313725490196,\"rgb(246,188,153)\"],[0.9411764705882353,\"rgb(247,212,187)\"],[1.0,\"rgb(250,234,220)\"]]},\"colorway\":[\"rgb(76,114,176)\",\"rgb(221,132,82)\",\"rgb(85,168,104)\",\"rgb(196,78,82)\",\"rgb(129,114,179)\",\"rgb(147,120,96)\",\"rgb(218,139,195)\",\"rgb(140,140,140)\",\"rgb(204,185,116)\",\"rgb(100,181,205)\"],\"font\":{\"color\":\"rgb(36,36,36)\"},\"geo\":{\"bgcolor\":\"white\",\"lakecolor\":\"white\",\"landcolor\":\"rgb(234,234,242)\",\"showlakes\":true,\"showland\":true,\"subunitcolor\":\"white\"},\"hoverlabel\":{\"align\":\"left\"},\"hovermode\":\"closest\",\"paper_bgcolor\":\"white\",\"plot_bgcolor\":\"rgb(234,234,242)\",\"polar\":{\"angularaxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"showgrid\":true,\"ticks\":\"\"},\"bgcolor\":\"rgb(234,234,242)\",\"radialaxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"showgrid\":true,\"ticks\":\"\"}},\"scene\":{\"xaxis\":{\"backgroundcolor\":\"rgb(234,234,242)\",\"gridcolor\":\"white\",\"gridwidth\":2,\"linecolor\":\"white\",\"showbackground\":true,\"showgrid\":true,\"ticks\":\"\",\"zerolinecolor\":\"white\"},\"yaxis\":{\"backgroundcolor\":\"rgb(234,234,242)\",\"gridcolor\":\"white\",\"gridwidth\":2,\"linecolor\":\"white\",\"showbackground\":true,\"showgrid\":true,\"ticks\":\"\",\"zerolinecolor\":\"white\"},\"zaxis\":{\"backgroundcolor\":\"rgb(234,234,242)\",\"gridcolor\":\"white\",\"gridwidth\":2,\"linecolor\":\"white\",\"showbackground\":true,\"showgrid\":true,\"ticks\":\"\",\"zerolinecolor\":\"white\"}},\"shapedefaults\":{\"fillcolor\":\"rgb(67,103,167)\",\"line\":{\"width\":0},\"opacity\":0.5},\"ternary\":{\"aaxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"showgrid\":true,\"ticks\":\"\"},\"baxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"showgrid\":true,\"ticks\":\"\"},\"bgcolor\":\"rgb(234,234,242)\",\"caxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"showgrid\":true,\"ticks\":\"\"}},\"xaxis\":{\"automargin\":true,\"gridcolor\":\"white\",\"linecolor\":\"white\",\"showgrid\":true,\"ticks\":\"\",\"title\":{\"standoff\":15},\"zerolinecolor\":\"white\"},\"yaxis\":{\"automargin\":true,\"gridcolor\":\"white\",\"linecolor\":\"white\",\"showgrid\":true,\"ticks\":\"\",\"title\":{\"standoff\":15},\"zerolinecolor\":\"white\"}}},\"xaxis\":{\"anchor\":\"y\",\"domain\":[0.0,1.0],\"scaleanchor\":\"y\",\"constrain\":\"domain\"},\"yaxis\":{\"anchor\":\"x\",\"domain\":[0.0,1.0],\"autorange\":\"reversed\",\"constrain\":\"domain\"},\"coloraxis\":{\"colorscale\":[[0.0,\"rgb(2,4,25)\"],[0.0625,\"rgb(24,15,41)\"],[0.125,\"rgb(47,23,57)\"],[0.1875,\"rgb(71,28,72)\"],[0.25,\"rgb(97,30,82)\"],[0.3125,\"rgb(123,30,89)\"],[0.375,\"rgb(150,27,91)\"],[0.4375,\"rgb(177,22,88)\"],[0.5,\"rgb(203,26,79)\"],[0.5625,\"rgb(223,47,67)\"],[0.625,\"rgb(236,76,61)\"],[0.6875,\"rgb(242,107,73)\"],[0.75,\"rgb(244,135,95)\"],[0.8125,\"rgb(245,162,122)\"],[0.875,\"rgb(246,188,153)\"],[0.9375,\"rgb(247,212,187)\"],[1.0,\"rgb(250,234,220)\"]]},\"title\":{\"text\":\"text-embedding-ada-002, at cutoff=0.8\"},\"height\":500,\"width\":1000},                        {\"responsive\": true}                    ).then(function(){\n",
       "                            \n",
       "var gd = document.getElementById('1b200369-dedd-401d-9410-e1359ddca891');\n",
       "var x = new MutationObserver(function (mutations, observer) {{\n",
       "        var display = window.getComputedStyle(gd).display;\n",
       "        if (!display || display === 'none') {{\n",
       "            console.log([gd, 'removed!']);\n",
       "            Plotly.purge(gd);\n",
       "            observer.disconnect();\n",
       "        }}\n",
       "}});\n",
       "\n",
       "// Listen for the removal of the full notebook cells\n",
       "var notebookContainer = gd.closest('#notebook-container');\n",
       "if (notebookContainer) {{\n",
       "    x.observe(notebookContainer, {childList: true});\n",
       "}}\n",
       "\n",
       "// Listen for the clearing of the current output cell\n",
       "var outputEl = gd.closest('.output');\n",
       "if (outputEl) {{\n",
       "    x.observe(outputEl, {childList: true});\n",
       "}}\n",
       "\n",
       "                        })                };                });            </script>        </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cutoff = 0.8\n",
    "fig = px.imshow(\n",
    "    torch.clamp(similarity_matrix, cutoff, 1),\n",
    "    title = f\"{embedder.model}, at cutoff={cutoff}\",\n",
    "    template='seaborn'\n",
    ")\n",
    "fig.update_layout(\n",
    "    height=500,\n",
    "    width=1000,\n",
    ")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "fd48c4ca-986a-4c1e-805f-4759c09d005e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 **Note:** This preprint has been reviewed by subject experts for *Review Commons*. Content has not been altered except for formatting.\n",
      "1 In this manuscript, Mishima et al., designed a reporter system (dubbed PACE, for Parallel Analysis of Codon Effects) to assess the effect of codon usage in regulating mRNA stability in a controlled sequence context. This reporter corresponds to a stretch of 20 repetitions of a given codon (to be tested for its effect on mRNA stability), each repetition being separated by one codon corresponding to each of the 20 canonical amino acids. This stretch is inserted at the 3\\' end of the coding sequence of a superfolder GFP flanked with fixed 5\\' and 3\\' untranslated regions. In vitro transcribed capped and polyadenylated RNAs are then produced from these reporters (each with a specific stretch of repetitions of a given codon), pooled together and injected into zebrafish zygotes to monitor their relative abundance at different time points upon injection.\n",
      "2 Using the PACE reporter, the authors were able to obtain a quantitative estimation of the impact of 58 out of the 61 sense codons on modulating mRNA stability. Their results are in agreement with a previous report that estimated the effect of codon usage on mRNA stability using endogenous mRNAs and an ORFeome library (Bazzini et al., 2016). However, contrary to relying on endogenous mRNAs and ORFeome reporters, the advantage of the PACE strategy is that the effect of the codon to be studied can be probed in a defined context, thus avoiding the presence of other motifs or transcript features that could also regulate mRNA stability. Similarly to results from Bazzini et al., 2016, the authors show that blocking translation completely abrogates the effect of codon usage, indicating that translation is required to drive codon-dependent mRNA degradation from their reporters. Also, the extent of codon-dependent mRNA decay is correlated with tRNA abundance and occurs through a process involving mRNA deadenylation as previously described in the zebrafish (Mishima et al., 2016 and Bazzini et al., 2016).\n",
      "3 Having validated their PACE protocol, the authors performed ribosome profiling to test whether ribosome occupancy on tested codons is correlated with their capacity to drive mRNA degradation. Their results indicate that, at least for polar amino acids, there is indeed an inverse correlation between ribosome occupancy at tested codons and mRNA stability thus suggesting that slow decoding of codons due to low levels of available cognate tRNA can induce mRNA degradation. The authors further validate this finding by reducing the levels of aminoacylated tRNAAsn (corresponding a polar amino acid) and showing that stability of the reporter RNA carrying a stretch of AAC codons (decoded by tRNAAsnGUU) is reduced. To test whether codon-dependent mRNA degradation in the context of slow ribosome decoding lead to ribosome stalling and collisions, the authors generated a mutant zebrafish strain with impaired expression of ZNF598 (an essential factor of the No-Go decay (NGD) pathway in yeast). They also integrated a known ribosome stalling sequence from hCMV (and a mutant version that does not trigger ribosome stalling) in their sfGFP reporter construct as a positive control for NGD in their assays. Their results indicate that although ZNF598 depletion impairs degradation of the hCMV reporter (as expected), it does not affect codon-dependent mRNA degradation, which appears to occur for most codons through a NGD-independent manner. Finally, through the use of a tandem ORF reporter assay separated by codon tags to be tested, the authors show that destabilizing codons do not stall ribosomes but only lead to their transient slowdown which induces mRNA deadenylation and degradation in a ZNF598-independent manner.\n",
      "4 Overall, the manuscript is very well written and pleasant to read. The introduction is well documented and relevant to the study as it allows readers to place the study in the current context of the field while highlighting open questions that have not been addressed yet. The results are clearly presented, the technical approaches are elegant and the conclusions convincing.\n",
      "5 Below you will find some major and minor points that, in my opinion, should be addressed by the authors.\n",
      "6 - One interesting aspect of the PACE reporter assay is the possibility to monitor ribosome occupancy in parallel for all codon-tags tested, which the authors did in Figure 3. However, instead of using RNA-seq data to normalize ribosome footprints and obtain ribosome occupancy, the authors used an alternative normalization approach consisting, for each codon-tag, to calculate the number of ribosome footprints with test codons in the A site divided by the number of ribosome footprints with spacer codons in the A site. This approach is elegant and appears to work with codons corresponding to polar amino acids. However, it might have its limitations for other codons.\n",
      "7 Indeed, ribosome dwell times (in yeast and mammals) have been shown to respond both to tRNA availability but also to other features such as the nature of the pair of adjacent codons, and the nature of the amino acid within the exit channel (Gobet C et al., 2020 PNAS; Gamble CE et al., 2016 Cell; Pavlov MY et al., 2009 PNAS). However, based on the work of 'Buschauer R et al., 2020 Science', only ribosomes lacking an accommodated tRNA at the A site are able to recruit Ccr4-Not to mediate mRNA deadenylation and degradation. Other events that increase ribosome dwell time (and thus occupancy), such as slow peptidyl-transfer, do not lead to Ccr4-Not recruitment and are resolved by eIF5A. It is therefore possible that depending on the nature of the codon that is being tested, ribosome occupancy at test and spacer codons can be biased by the nature of codon-pairs and 'dilute' the effects of tRNA availability.\n",
      "8 If the authors performed RNA-seq together with the ribosome profiling experiment, it might be interesting to use the RNA-seq data to calculate ribosome occupancy on 'tested' and 'spacer' codons to check whether using this normalization, they do find a negative correlation between ribosome occupancy and PACE stability. A different approach would be to perform ribosome run-off experiments using harringtonine and estimate the elongation speed across the codon tag. However, I am aware that this experiment could be tedious an expensive.\n",
      "9 - Figure 6: Insertion of the Lys x8 AAA stretch in the tandem ORF reporter leads to a decrease in HA-DsRedEx expression compared to that of Myc-EGFP. However, results from 'Juszkiewicz and Hedge, 2017' using a similar reporter in mammalian cells indicate that stretches of Lys AAA below 20 repetitions only elicit poor RQC (less than 10% of true ribosome stalling for 12 repetitions of the AAA codon). Instead, most of the loss in RFP signal results from a change in the reading frame of ribosomes due to the 'slippery' translation of the poly(A) stretch. I therefore think that it could be important to perform the experiment in ZNF598 KO embryos to validate that the observed reduction in HA-dsRedEx does indeed result from stalling and RQC and not from a change in the reading frame of ribosomes.\n",
      "10 On a similar note, how do the authors explain the decrease in signal of the Flag-EGFP and HA-DsRedEx observed when using the Flag-EGFP with non-optimal codons? I understand that RQC occurring through NGD leads to ribosome disassembly at the stalling site and possibly mRNA cleavage (thus explaining the decrease in HA-DsRedEx signal compared to Myc-EGFP). However, I would assume that codon-mediated mRNA decay (even for ORF longer than 200 of non-optimal codons) should trigger mRNA deadenylation, followed by decapping and co-translational 5\\'to3\\' mRNA degradation, following the last translating ribosome. I would therefore expect not to see any change in the HA-DsRedEx/Myc-EGFP ratio even for the non-optimal Flag-EGFP reporter. Could the 200 non-optimal codons trigger some background RQC through NGD? Or could there be some ribosome drop-off? It might be interesting to test the optimal and non-optimal Flag-EGFP reporters in the ZNF598 KO background to check whether the observed decrease in the relative amount of HA-DsRedEx results from stalling-dependent RQC.\n",
      "11 - The color-coded CSC results from 'Bazzini et al., 2016' presented at the bottom of panel B in figure 2 are misleading because many codons (such as PheUUU, AsnAAU, TyrUAC...) are lacking information. I have the impression that the authors used the combined data from the rCSCI (obtained from the reporter RNAs) and CSC (obtained from endogenous transcripts) corresponding to Figure 1F from Bazzini et al., 2016. This data set excluded all codons that were not concordant between the endogenous and reporter CSCs (which are those that are lacking a color code in Figure 2B from this study). However, in the scatter-plot of PACE Vs CSC (from Supplemental Figure 1D of this study), the authors used the complete set of CSC values from Bazzini et al .,2016. Could the authors please use the complete set of CSC values from Bazzini et al., 2016 to color code codons in their Figure 2B?\n",
      "12 - Figure 4B. The charged tRNA measurements seem to have been done in a single biological replicate (there aren\\'t any error bars in the chart). I understand that the procedure is tedious and requires a large amount of total RNA to begin with, but it would be preferable to perform it in three biological replicates.\n",
      "13 - Supplementary Figure 2B. I do not understand what the figure represents. The legend is quite cryptic and states that the panel corresponds to the information content of each reading frame. More information should be given so that readers can understand how to interpret de figure and extract periodicity information.\n",
      "14 Since the seminal work from Jeff Coller\\'s laboratory in 2015 (Presnyak et al., 2015 Cell) showing a global and major role for codon optimality in determining mRNA half-lives in yeast, the role of codon usage in modulating translation and stability of mRNAs has been widely studied in different organisms (including zebrafish and mammals). As stated by the authors in the introduction, most studies have relied on correlation analyses between codon usage and mRNA half-lives from endogenous transcripts or from ORF libraries with fixed 5\\'UTR and 3\\'UTRs. This approaches could suffer from the presence of transcript features that can participate in other mRNA degradation pathways, which could limit their use when performing further mechanistic studies.\n",
      "15 The work by Mishima and collaborators presents an original reporter assay that allows to evaluate the role of codon usage on regulating mRNA stability in a defined context, thus avoiding the impact of confounding factors that could bias the measurement of mRNA stability. Results obtained using this reporter are in good agreement with previous reports from Zebrafish (Bazzini et al 2016., and Mishima et al., 2016). From this validated reporter approach, the authors further show that codon-dependent mRNA degradation is directly related to tRNA availability and (at least partially) to ribosome occupancy (two factors already suggested as being important for codon-mediated decay in zebrafish, although they were based on correlation analyses).  Furthermore, the authors show that codon-mediated mRNA decay occurs during productive mRNA translation and that it is functionally distinct from RQC induced by ribosome stalling. As a consequence, codon-mediated mRNA degradation is independent from the RQC factor ZNF598 (which they also validate for the first time as an important RQC factor in zebrafish). This information is new within metazoans since only in yeast it has been clearly shown that codon-mediated mRNA decay is distinct from RQC induced by ribosome stalling and collisions.\n",
      "16 Taken together, the reported findings will be of interest to the community working on mRNA metabolism and translation. It could also interest, more broadly,  scientists working on translational selection and genome evolution.\n"
     ]
    }
   ],
   "source": [
    "for i, p in enumerate(review_chunks):\n",
    "    print(i, p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "48964302-24c4-4e86-88b0-18faca90b95e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 During the translation of mRNA to protein, a sequence of codons is decoded by tRNAs on the ribosome. While the translation elongation process is strictly controlled, the speed of ribosome movement passing through codons is not uniform. For example, each codon is decoded with variable efficiency due to the difference in tRNA availability (Hussmann et al., 2015; Varenne et al., 1984; Weinberg et al., 2016). Particular pairs of amino acids and codons inhibit peptide bond formation and/or decoding (Doerfel et al., 2013; Gamble et al., 2016; Schuller et al., 2017; Ude et al., 2013). The nascent polypeptide may interact with the ribosome exit tunnel and retard the ribosome (Ito and Chiba, 2013; Wilson et al., 2016). As a result, ribosomes traverse along the ORF with distinct kinetics unique to each coding sequence (Choi et al., 2018).\n",
      "1 Nonuniform movement of the ribosome is not always harmful and is instead beneficial to the cell (Stein and Frydman, 2019). In bacteria, programmed ribosome stall by specific nascent peptides acts as a regulatory switch to control protein expression (Ito and Chiba, 2013; Wilson et al., 2016). Ribosome slowdown provides duration time for nascent chain folding and complex formation (Buhr et al., 2016; Pechmann et al., 2014; Pechmann and Frydman, 2013; Stein et al., 2019; Weinberg et al., 2016). The 5’ end of the ORF tends to contain a stretch of uncommon codons that is decoded inefficiently and optimizes the ribosome traffic right downstream of the initiation codon (Tuller et al., 2010; Weinberg et al., 2016). Hence, cells harness the nonuniform translation elongation process to control protein output.\n",
      "2 In addition to proteome quality, ribosome movement impacts the stability of mRNAs. Transcriptome-wide analysis of mRNA half-life in budding yeast revealed a correlation between codon frequency and mRNA stability: some codons are enriched in stable mRNAs, whereas others are enriched in unstable mRNAs (Presnyak et al., 2015). Similarly, the stability of thousands of maternal mRNAs in zebrafish embryos is attributable to their codon composition (Bazzini et al., 2016; Mishima and Tomari, 2016). A correlation between codons and mRNA stability has been reported in multiple organisms based on analyses of endogenous mRNAs and ORFenome reporter mRNAs, indicating a conserved role of codons in determining mRNA stability (hereafter called codon-mediated decay) (Bazzini et al., 2016; Boël et al., 2016; Forrest et al., 2020; Harigaya and Parker, 2016; Hia et al., 2019; Narula et al., 2019; Wu et al., 2019).\n",
      "3 As codon-mediated decay is dependent on translation, it is the ribosome that connects codons to mRNA stability (Bazzini et al., 2016; Forrest et al., 2020; Mishima and Tomari, 2016; Narula et al., 2019; Presnyak et al., 2015; Wu et al., 2019). Indeed, codon effects on mRNA stability negatively correlate with tRNA availability and ribosome density relative to A-site codons in yeast (Hanson et al., 2018). These correlations led to a model wherein codon optimality determines mRNA stability: codons that are slowly decoded due to low tRNA availability confer an mRNA-destabilizing effect, whereas codons with high tRNA availability are decoded smoothly and confer an mRNA-stabilizing effect (Presnyak et al., 2015). A subsequent study in yeast revealed that the recruitment of Not5 to the E site of the slowed ribosome and ubiquitination of ribosomal protein eS7 by Not4 are both crucial for codon-mediated decay (Buschauer et al., 2020). As Not5 and Not4 are components of the Ccr4-Not deadenylase complex, these mechanisms explain how slow ribosomes promote mRNA deadenylation and decay.\n",
      "4 Codon-mediated decay accompanies mRNA deadenylation not only in yeast but also in zebrafish and humans (Bazzini et al., 2016; Mishima and Tomari, 2016; Presnyak et al., 2015; Webster et al., 2018; Wu et al., 2019). Furthermore, recent analyses in human cells revealed that the codon stability coefficient (CSC), ribosome density relative to the A site, and tRNA availability are correlated with each other (Forrest et al., 2020; Narula et al., 2019; Wu et al., 2019). Although tRNA availability is a complex parameter influenced by multiple factors, such as tRNA amount, tRNA aminoacylation level, and amino acid amount, these studies suggested conservation of the molecular mechanism underlying codon-mediated decay. However, genome-wide analyses and ORFeome rely on a mere correlation between mRNA stability and codon composition on a genome-wide scale and therefore lack a direct and experimental measurement of each single codon effect on mRNA stability (Bazzini et al., 2016; Forrest et al., 2020; Hanson et al., 2018; Narula et al., 2019; Presnyak et al., 2015; Wu et al., 2019).\n",
      "5 Moreover, earlier works may include effects from related but distinct mRNA decay pathways that also act via ORFs. One such example is a translation quality control pathway that monitors the elongating ribosomes. Slowed ribosomes often induce collisions with trailing ribosomes and form specific structures called disome and trisome, a subset of which is recognized by the E3 ubiquitin ligase Hel2 in yeast. Hel2 ubiquitinates specific ribosomal proteins in the 40S subunit to initiate ribosome-associated quality control (RQC) and endonucleolytic cleavage of mRNAs via no-go decay (NGD) (Inada, 2020; Joazeiro, 2019). RQC is conserved in mammals, in which the Hel2 ortholog Znf598 acts as a collision sensor (Garzia et al., 2017; Juszkiewicz et al., 2018; Juszkiewicz and Hegde, 2017; Sundaramoorthy et al., 2017). In contrast, the occurance of NGD is controversial in vertebrates: the established RQC substrate, consecutive Lys AAA codons, does not promote apparent mRNA decay in vertebrates (Juszkiewicz and Hegde, 2017), whereas recruitment of the Znf598-4EHP-GIGYF1/2 complex on paused ribosomes promotes mRNA decay (Weber et al., 2020). Hence, it remains unclear whether the codon effects captured by genome-wide analyses in vertebrates are entirely attributable to codon-mediated decay since they may include the effect of NGD-related pathways. A simplified experimental approach that recapitulates the essence of codon-mediated decay should complement genome-wide analysis and better define the ribosome status that triggers cotranslational mRNA decay pathways.\n",
      "6 Here, we analyze codon-mediated decay in zebrafish embryos using an artificial yet simplified reporter system. Our approach captures each single codon effect on mRNA stability in a defined context. Using this system, we evaluated the correlations between codon effects, ribosome occupancy, and the tRNA amount. We also characterized the mRNA decay pathway induced by ribosome stalls in zebrafish and analyzed its relationship to codon-mediated decay. Finally, we compared the speed of the ribosome translating mRNA-destabilizing codons to that of an aberrantly stalled ribosome. Our data provide a better definition of codon-mediated decay in zebrafish as a molecular event that occurs during slow but productive translation elongation cycles.\n",
      "7 We designed an artificial codon-tag sequence to analyze the codon effects on mRNA stability in a defined sequence context (Figure 1A). This sequence contains a codon to be tested (e.g., a codon for Leu) 20 times, each separated by a codon encoding one of the 20 amino acids. The sequence was inserted at the 3’ end of the superfolder GFP (sfGFP) ORF by taking the positional effect of codon-mediated decay into account (Mishima and Tomari, 2016). We then compared the CUG and CUA Leu codon tags, whose differential effects on deadenylation were reported previously in zebrafish embryos (Mishima and Tomari, 2016). In vitro synthesized, capped and polyadenylated reporter mRNAs were injected into 1-cell stage zebrafish embryos and analyzed at two hours post fertilization (hpf) (before the maternal-to-zygotic transition [MZT]) and 6 hpf (after the MZT). Analysis of the poly(A) tail lengths by PAT assays revealed that the CUA codon-tag promoted deadenylation compared to the CUG codon-tag (Figure 1B and 1C). Consistent with the poly(A) tail status, qRT-PCR showed that the CUA codon-tag reporter was less stable than the CUG codon-tag reporter (Figure 1D). The observed difference in mRNA stability was a cotranslational effect because inhibition of translation initiation by an antisense morpholino oligonucleotide (MO) specific to the GFP ORF abolished the destabilizing effect of the CUA codon (Figure 1E). Hence, our codon-tag reporters recapitulated the translation-dependent effect of codons on mRNA deadenylation and degradation in zebrafish embryos.\n",
      "8 (A) A scheme of the codon-tag reporter system. A codon to be tested (pink, in this case leucine) and one of the 20 sense codons (gray) are alternately repeated 20 times. This codon-tag sequence was inserted at the end of the sfGFP ORF. (B) Poly(A) tail analysis of codon tag reporter mRNAs at 2 and 6 hpf. The developmental stages are shown above. The lane labeled A0 shows the 3’ UTR fragment without a poly(A) tail. (C) Quantification of the PAT assay in (B). (D and E) qRT-PCR analysis of codon tag reporter mRNA levels at 6 hpf relative to 2 hpf in the absence (D) or presence (E) of translation-blocking GFP MO. Values are means ± SD of three independent experiments. P-values were calculated using two-sided Student’s t-test.\n",
      "9 Having validated the codon-tag reporter system with the two Leu codons, we constructed a library of codon-tag reporters to analyze multiple codon effects in parallel (hereafter called parallel analysis of codon effects; PACE). We prepared a mixture of codon-tag reporter mRNAs by plasmid cloning and in vitro transcription using this approach. We then injected a mixture of reporter mRNAs into zebrafish embryos and performed RNA sequencing (RNA-Seq) at two different time points (2 hpf and 6 hpf) to quantify the reads mapped on each tag sequence (Figure 2A). Codon-tag reads were normalized to the total reads mapped to the zebrafish genome, and the relative codon-tag amount at 6 hpf compared to 2 hpf was calculated as the effect of each codon on mRNA stability. Although four codon tags were not included due to technical limitations in library preparation and/or sequencing (UCU, UCA, GGG, and a stop codon UAG), this approach allowed us to compare the effect of 58 sense codons on mRNA stability in a defined sequence context with reasonable reproducibility (Figure 2B and S1A).\n",
      "10 (A) Scheme of parallel analysis of codon effects (PACE). Codon tag reporter mRNAs were pooled as a library and coinjected into fertilized zebrafish eggs. Embryos were collected at 2 and 6 hpf and subjected to RNA-Seq analysis. Reads mapped on each codon tag region were quantified to calculate each codon’s effect on mRNA stability. (B) Results of PACE in zebrafish embryos. Black circles show the relative stability of reporter mRNAs under normal translation. The stability of a codon-tag reporter with a UGA stop codon is set to one. Gray circles show the relative stability of reporter mRNAs under the condition in which translation initiation was inhibited by GFP MO. Averages of two biological replicates are shown. The relative effect of each codon on mRNA stability is shown as a color gradation with red (strong) to blue (weak). Codon effects estimated in Bazzini et al. 2016 are indicated below as red (nonoptimal-strong), light red (nonoptimal-weak), light blue (optimal-weak), and blue (optimal-strong) circles. (C) PACE results shown in (B) are presented as a codon table. (D) Box plot of tRNA amounts for nonoptimal (red), neutral (light gray), and optimal (blue) codons based on CSC (Bazzini et al., 2016). (E) Box plot of tRNA amounts for codons with strong (red), neutral (light gray), and weak (red) destabilizing effects based on PACE. The box represents the interquartile range (IQR), with the median indicated as a black horizontal line in the box. The whiskers represent the variation within 1.5 IQR outside the upper and lower quartiles. Outliers are shown as dots.\n",
      "11 Several lines of evidence indicated that PACE was a valid approach for the analysis of codon-mediated decay. First, PACE detected codon-specific effects on mRNA stability. The effects were variable between synonymous codons; for example, among three synonymous codons for Ile, AUU and AUA showed a more substantial destabilizing effect than AUC (Figure 2B and 2C). A similar difference was observed in most codon boxes, although some two-codon boxes (Tyr, Cys, Asp, and Glu) and the Gly codon box showed limited variability. Second, the variable effects were dependent on translation. Inhibition of translation initiation by GFP MO stabilized almost all codon-tag reporters to a similar level, thereby reducing the difference between codons (Figure 2B and S1B). Third, PACE showed good correlations with codon frequency in the zebrafish genome (Pearson correlation coefficient r = 0.519; Figure S1C) and CSC (r = 0.689, Figure S1D), both of which were used to infer the codon effects on mRNA stability in previous genome-wide studies (Bazzini et al., 2016; Mishima and Tomari, 2016). In particular, codons classified as optimal or nonoptimal based on CSC were clearly separated in PACE (Figure 2B). We also confirmed the accuracy of PACE by qRT-PCR analysis of individual reporter mRNAs (Figure S1E and S1F).\n",
      "12 (A) A scatter plot showing the reproducibility of the codon effects measured by PACE in two replicates. (B) A scatter plot comparing the codon effects measured by PACE in the absence (x-axis) and presence of GFP MO (y-axis). (C) A scatter plot comparing codon usage in the zebrafish genome (per thousand codons, x-axis) and the codon effects measured by PACE (y-axis). (D) A scatter plot comparing CSCs (Bazzini et al., 2016, x-axis) and the codon effects measured by PACE (y-axis). (E and F) Comparison of the codon effects measured by PACE (E) with the results of qRT-PCR (F). mRNA levels at 6 hpf relative to 2 hpf are shown. The experiments were repeated three times. Error bars show SD. In A-D, the regression line is shown in blue and the 95% confidence interval is shown in gray. r represents Pearson’s correlation. The significance was calculated by Student’s t-test.\n",
      "13 Using CSC and tRNA sequencing, Bazzini et al. showed that nonoptimal codons tend to have fewer corresponding tRNAs than optimal codons in zebrafish (Bazzini et al., 2016) (Figure 2D). This trend was also true with the codon effects measured by PACE. When we divided codons into three groups based on the destabilizing effects in PACE (strong, neutral, and weak), we observed that codons with strong destabilizing effects tended to have fewer corresponding tRNAs (Figure 2E). Overall, we concluded that PACE successfully captured cotranslational, codon-driven effects on mRNA stability in zebrafish embryos.\n",
      "14 Next, we asked if the codon effects measured by PACE correlated with the ribosome elongation process. To this end, we performed ribosome footprint profiling with PACE reporters in zebrafish embryos (Han et al., 2020; Ingolia et al., 2011, 2009). We collected embryos injected with the PACE library at 4 hpf, a middle time point between 2 hpf and 6 hpf, to capture ribosome movement during the process of codon-mediated decay. Embryos were snap-frozen and treated with cycloheximide after cell lysis, which avoids aberrant effects of the compound on ribosome occupancy at codons (Hussmann et al., 2015; Weinberg et al., 2016). The obtained reads fell into 28-30 nt lengths, a hallmark of ribosome footprints (Figure S2A). Robust three-nucleotide periodicity was ensured by the information contents in the 28-30 nt fragments (Figure S2B). To analyze ribosome occupancy on codon tags, we counted the footprints with the A-site position placed on the given test codon and then normalized that by the footprints with the A site at the spacer codons (Figure 3A). This calculation allowed us to estimate the density of ribosomes waiting for decoding test codons without a complex normalization procedure. The numbers of footprints were well correlated in two replicates (Figure S2C and S2D); thus, we combined reads from two replicates to calculate the ribosome density.\n",
      "15 (A) Distribution of the ribosome footprint lengths derived from codon-tag sequences. The results of the two replicates are plotted separately in red and blue. (B) Information content of each reading frame using ribosome footprints aligned to the codon tag reporter sequences. The results of the two replicates are plotted separately in red and blue. (C and D) Scatter plots showing the reproducibility of ribosome footprint analysis with PACE reporters. Two replicates for test codon reads (C) and spacer codon reads (D) are shown. The regression line is shown in blue and the 95% confidence interval is shown in gray. r, Pearson’s correlation. Significance was calculated by Student’s t-test.\n",
      "16 (A) Scheme of ribosome footprint analysis using PACE reporter mRNAs. An example of a 28 nt footprint is shown. The relative ribosome density of each codon at the A site was calculated by the formula presented on the right. (B-E) Scatter plots showing a correlation between relative ribosome density (28-30 nt footprints; x-axis) and the codon effects measured by PACE (y-axis). Each dot represents a single codon. Graphs for all codons (B), codons for polar amino acids (C), nonpolar amino acids (D), and charged amino acids (E) are shown. (F-I) Scatter plots showing a correlation between relative ribosome density (28-30 nt footprints; x-axis) and the tRNA amounts (y-axis, Bazzini et al., 2016). Each dot represents a single codon. Graphs for all codons (F), codons for polar amino acids (G), nonpolar amino acids (H), and charged amino acids (I) are shown. In B-I, the regression line is shown in blue and the 95% confidence interval is shown in gray. P-values were calculated using Student’s t-test.\n",
      "17 We first analyzed a correlation between the codon effects measured by PACE and the ribosome density for all codons tested, but no clear correlation was observed (Figure 3B). Because translation of a codon tag produces a nascent peptide enriched with a particular amino acid locally, the nascent peptide within the ribosome exit tunnel may impact on the speed of ribosome traverse. Therefore, we divided the codon tags into three groups based on the characteristics of the encoded amino acid: polar, nonpolar, and charged. This classification revealed that codon tags encoding polar amino acids showed a negative correlation between the codon effects and the ribosome density: the higher the ribosome density on a given codon, the stronger the destabilization effect (Figure 3C). On the other hand, such a correlation was not observed with codons for nonpolar and charged amino acids (Figure 3D and 3E). We compared the ribosome density of tested codons to the corresponding tRNA amounts to further analyze the relationship between the ribosome density and the decoding process. Consistent with the analysis of ribosome density and the PACE results, a negative correlation between ribosome density and tRNA amounts was observed with codons for polar amino acids (Figure 3F-I). Although there was a technical limitation (see discussion), our PACE approach confirmed that codon-mediated decay is connected to the slower codon-decoding process at the A site by tRNA, at least for polar amino acid codons.\n",
      "18 Next, we modulated tRNA availability in zebrafish embryos to experimentally validate the connection between codon effects and the decoding process. The bacterial enzyme AnsB, which hydrolyzes asparagine (Asn) to aspartate (Asp), reduces aminoacylated tRNAAsn and increases the duration of the ribosome on Asn codons in human cells (Loayza-Puch et al., 2016). Overexpression of AnsB in zebrafish embryos by mRNA injection caused morphological defecst at the shield stage (Figure 4A). Analysis of charged tRNA levels in these embryos by OXOPAP (Gaston et al., 2008) followed by qRT-PCR revealed that aminoacylated tRNAAsnGUU was specifically reduced in the presence of AnsB (Figure 4B). We then examined whether the stable Asn AAC codon tag reporter mRNA was subject to codon-mediated decay in the presence of AnsB. We found that the AAC codon tag reporter mRNA was less stable in AnsB-expressing embryos than in control embryos (Figure 4C). In contrast, the stability of the unrelated codon tag reporter mRNA for Leu CUG was not affected by AnsB. The PAT assay revealed that deadenylation of the AAC codon tag reporter mRNA was promoted in AnsB-expressing embryos (Figure 4D and 4E), whereas the poly(A) tail status of the CUG codon tag reporter mRNA was not affected (Figure 4F and 4G). Hence, experimental modulation of a particular tRNA availability (tRNAAsnGUU) altered the effect of the corresponding codon (AAC, which is decoded by tRNAAsnGUU). These results support the idea that tRNA availability is a major determinant of codon-mediated deadenylation and mRNA degradation in zebrafish embryos.\n",
      "19 (A) AnsB causes asparagine deprivation by converting asparagine to aspartic acid. Mock-injected and AnsB-overexpressing zebrafish embryos at 6 hpf are shown. (B) Charged tRNA levels in mock-injected and AnsB-overexpressing embryos at 6 hpf. Total and charged tRNAs were measured by qRT-PCR coupled to the OXOPAP assay. The relative charged tRNA level in mock-injected embryos was set to one. (C) qRT-PCR analysis of Asn AAC and Leu CUG codon tag reporter mRNAs at 6 hpf relative to 2 hpf in mock-injected and AnsB-overexpressing embryos. Experiments were repeated three times, and the average is shown. Error bars show SD. P-values were calculated using two-sided Student’s t-test. (D) Poly(A) tail analysis of Asn AAC codon tag reporter mRNA at 2 and 6 hpf. The developmental stages are shown above as hpf. The lane labeled A0 shows the 3’ UTR fragment without a poly(A) tail. (E)) Quantification of the PAT assay in (D). (F) Poly(A) tail analysis of Leu CUG codon tag reporter mRNA at 2 and 6 hpf. (G) Quantification of the PAT assay in (F).\n",
      "20 In theory, the slowed ribosome at the decoding step may drive other types of cotranslational mRNA decay such as NGD. Thus, we tested whether the codon effects observed in our PACE experiments represented codon-mediated decay or a mixed output with NGD.\n",
      "21 As NGD has not been previously reported in zebrafish, we first had to investigate whether the ribosome stall induces NGD in zebrafish embryos. For this purpose, we focused on the ribosome stalling sequence from human cytomegalovirus (hCMV) uORF2, whose peptide arrests the ribosome on the stop codon together with eRF1 (Matheisl et al., 2015). Here, we hypothesized that hCMV uORF2 causes ribosome stalls to be strong enough to induce NGD in zebrafish embryos. When we inserted codon-optimized hCMV uORF2 at the 3’ end of the sfGFP ORF, protein output was significantly reduced in zebrafish embryos (Figure 5A and 5B). As reported, mutating the diproline motif at the end of uORF2, critical amino acids for ribosome stalling, partially restored the protein output (Matheisl et al., 2015). We then measured the stability of these reporter mRNAs and found that hCMV uORF2 reduced mRNA stability in a diproline motif-dependent manner (Figure 5C). Notably, the PAT assay detected no significant deadenylation of the reporter mRNAs (Figure 5D and E). These results suggested that hCMV uORF2 induced mRNA decay in a deadenylation-independent manner, distinct from codon-mediated decay.\n",
      "22 (A) A scheme of the hCMV uORF2 reporter mRNAs. Wild-type hCMV uORF2 (light blue) or its mutant (dark blue) was inserted at the end of the superfolder GFP (sfGFP) ORF. (B) Western blotting to detect the sfGFP protein levels at 6 hpf. Tubulin (Tub) was detected as a loading control. (C) qRT-PCR analysis of hCMV uORF2 reporter mRNA levels in wild-type embryos at 6 hpf relative to 2 hpf. The experiments were repeated three times. Error bars show SD. (D) Poly(A) tail analysis of hCMV uORF2 reporter mRNAs at 2 and 6 hpf. The developmental stages are shown above as hpf. The lane labeled A0 shows the 3’ UTR fragment without a poly(A) tail. (E)) Quantification of the PAT assay in (D). (F) Scheme of human and zebrafish Znf598 proteins. The sgRNA target site is indicated with an arrowhead. Genome sequences of wild-type and znf598 mutant zebrafish are shown below. The predicted amino acid sequence of mutated Znf598 is shown in red. The asterisk indicates the premature stop codon. (G) qRT-PCR analysis of hCMV uORF2 reporter mRNA levels in MZznf598 embryos at 6 hpf relative to 2 hpf. The experiments were repeated three times. Error bars show SD. (H) qRT-PCR analysis of hCMV uORF2 reporter mRNA levels in wild-type embryos, MZznf598 embryos and MZznf598 embryos injected with an mRNA encoding Myc-tagged full-length (full) or RING domain deleted (ΔRING) Znf598. mRNA levels at 6 hpf relative to 2 hpf are shown. The experiments were repeated three times. Error bars show SD. (I) Western blotting to detect the Myc-tagged Znf598 proteins at 6 hpf. Tubulin was detected as a loading control. (J) Results of PACE in MZznf598 zebrafish embryos. Black circles show the relative stability of the reporter mRNAs. The stability of a codon-tag reporter with a UGA stop codon is set to one. Averages of two biological replicates are shown. The relative effect of each codon on mRNA stability in wild-type embryos measured in Figure 2B is shown as a color gradation with red (strong) to blue (weak). In C, G, and H, P-values were calculated using two-sided Student’s t-test.\n",
      "23 RQC-coupled NGD in yeast requires the E3 ubiquitin ligase Hel2 and subsequent ubiquitination of ribosomal protein uS10 (Ikeuchi et al., 2019). To characterize mRNA decay induced by hCMV uORF2, we generated a zebrafish mutant for Znf598, a vertebrate homolog of Hel2, by using CRISPR-Cas9. We obtained a strain containing an 11 bp deletion in exon 1 of the znf598 locus, which was predicted to induce a frameshift and produce a protein product truncated within the RING finger domain (Figure 5F). Embryos lacking both maternal and zygotic Znf598 (MZznf598) were obtained by crossing homozygous znf598 mutants. znf598 mRNA was significantly reduced in MZznf598 (Figure S3A), likely reflecting the premature termination of translation followed by nonsense-mediated mRNA decay (Maquat, 2004). The detailed phenotype of our znf598 mutant will be described elsewhere. Analysis of the reporter mRNA stability revealed that hCMV uORF2 did not induce mRNA decay in MZznf598 embryos (Figure 5G). Rescue experiments revealed that the mRNA decay activity of hCMV uORF2 was restored by full-length Znf598 but not by a deletion construct lacking the RING domain (Figure 5H and I). These results showed that specific ribosome stalls induced mRNA decay in a Znf598-dependent manner in zebrafish embryos.\n",
      "24 (A) qRT-PCR analysis of actb1 and znf598 mRNA levels in wild-type and MZznf598 embryos at 24 hpf. The results are normalized with the 18S rRNA level. The mRNA level in wild type is set to one. The experiments were repeated three times. Error bars show SD. (B) A scatter plot comparing the codon effects measured by PACE in MZznf598 (x-axis) and wild-type (y-axis) embryos. The regression line is shown in blue and the 95% confidence interval is shown in gray. r, Pearson’s correlation. Significance was calculated by Student’s t-test.\n",
      "25 This observation led us to test the impact of Znf598 on codon effects measured by PACE (Figure 5J). Overall, we found that most codon effects were maintained in MZznf598 compared to wild type, with a high correspondence (the correlation coefficients 0.826 [Pearson] and 0.835 [Spearman]) (Figure 5J and S3B). However, we noted that a few codons changed their effects in MZznf598: Ala GCG turned into a destabilizing codon, whereas Lys AAA and Trp UGG codon tag reporter mRNAs were more stable in MZznf598 than wild type. Since ribosome stalls formed by consecutive AAA and UGG codons are subject to RQC in yeast and humans (Dimitrova et al., 2009; Juszkiewicz and Hegde, 2017; Mizuno et al., 2021; Sundaramoorthy et al., 2017), these codon tags might induce mRNA decay in part via the NGD-like pathway in zebrafish embryos. We draw two conclusions from these results: 1) Znf598-dependent NGD-like activity exists in zebrafish embryos and 2) most if not all of the codon effects measured by PACE are distinct from Znf598-dependent mRNA decay.\n",
      "26 To further characterize the nature of the ribosome slowdown that leads to codon-mediated decay, we performed a tandem ORF reporter assay that could quantitatively detect ribosome stalling in mammalian cells (Juszkiewicz and Hegde, 2017; Sundaramoorthy et al., 2017). Myc-tagged enhanced GFP (EGFP) and HA-tagged DsRed-Express (DsRedEx) were separated by two porcine teschovirus-1 2A (P2A) sequences that cause skipping of peptide bond formation (Kim et al., 2011), and a sequence to be tested for the efficiency of translation elongation was inserted between the two P2A sequences (Figure 6A). If the inserted sequence inhibited the elongation process of the ribosome, the DsRedEx/EGFP ratio would be reduced in this experimental setup. In vitro synthesized mRNA was injected into 1-cell stage zebrafish embryos, and the protein expression was analyzed at 3 hpf by western blotting to detect protein levels before the onset of codon-mediated decay.\n",
      "27 (A) A schematic of the tandem ORF assay. Myc-tagged EGFP ORF (green) and HA-tagged DsRedEx ORF (red) are separated by two P2A translation skipping sequences (light blue). A sequence to be tested (gray) is inserted between the two P2A sequences. (B) Results of the tandem ORF assay with Lys AAG×8 and Lys AAA×8. Representative western blotting detecting Myc-tagged EGFP and HA-tagged DsRedEx at 3 hpf is shown on the left. Quantification of the HA-DsRedEx levels normalized by Myc-EGFP is shown on the right. The HA-DsRedEx level with no insert sequence is set to one. (C) Results of the tandem ORF assay with Asn AAC and Asn AAU codon tags. Representative western blotting detecting Myc-tagged EGFP and HA-tagged DsRedEx at 3 hpf is shown on the left. Quantification of the HA-DsRedEx levels normalized by Myc-EGFP is shown on the right. The HA-DsRedEx level with no insert sequence is set to one. (D) Results of the tandem ORF assay using EGFP ORFs with different codon optimality. Representative western blotting detecting Myc-tagged EGFP, Flag-tagged codon-modified EGFP and HA-tagged DsRedEx at 3 hpf is shown on the left. Quantification of Flag-EGFP and HA-DsRedEx levels normalized by Myc-EGFP is shown on the right. The Flag-EGFP or HA-DsRedEx level with no insert sequence was set to one. All experiments were repeated three times, and the average is shown. Error bars show SD. P-values were calculated using two-sided Student’s t-test.\n",
      "28 To verify the sensitivity of the assay, we first analyzed the effect of the consecutive Lys AAA codons that are known to induce ribosome stall, collision, and RQC in other organisms (Ito-Harashima et al., 2007; Juszkiewicz and Hegde, 2017; Sundaramoorthy et al., 2017). Consistent with previous studies, ribosome stall at the eight consecutive AAA codons was detected as an approximately 40 % reduction in DsRedEx expression compared to the consecutive Lys AAG codons (Figure 6B). We then analyzed the effect of two synonymous AAC and AAU Asn codon tags, which showed opposite impacts on mRNA stability in PACE (Figure 2B and C). We found that Asn codon tag reporters showed almost identical DsRedEx/EGFP ratios (Figure 6C). This result suggested that ribosomes were not significantly retarded on the destabilizing codons. To confirm this idea with other codons, we compared two EGFP ORFs encoded by either optimal or nonoptimal codons throughout the ORF (Mishima and Tomari, 2016). The inserted nonoptimal EGFP ORF tended to reduce downstream DsRedEx expression, but the difference was quite subtle and statistically insignificant (Figure 6D). Hence, the translation of >200 nonoptimal codons was still more efficient than that of a short, stall-prone polyA stretch. These results showed that the ribosome slowdown required for inducing codon-mediated decay is far more transient than the ribosome stall required for inducing ribosome collision and RQC.\n",
      "29 In this study, we developed a reporter-based assay named PACE to analyze codon-mediated decay in zebrafish embryos. PACE was sensitive enough to detect translation-dependent, codon-specific effects on mRNA stability in zebrafish embryos (Figure 2). PACE relies on relatively simple bioinformatics analysis and is potentially applicable to other organisms capable of undergoing mRNA injection or transfection. Optimization of the codon tag sequence and the inclusion of the four codon tags missed during its construction or RNA-seq will broaden the utility of PACE.\n",
      "30 Previous studies estimated codon effects by calculating the correlation between the codon frequency and mRNA stability for thousands of endogenous mRNAs (Bazzini et al., 2016; Forrest et al., 2020; Harigaya and Parker, 2016; Mishima and Tomari, 2016; Presnyak et al., 2015). These analyses took advantage of averaging codon effects in multiple different positions on the transcriptome. However, the correlation tended to be modest in part due to the presence of factors that affect mRNA stability independent of codons (e.g., RNA-binding proteins and microRNAs) (Medina-Muñoz et al., 2021). The ORFome approach minimizes this disadvantage by fixing the 5’ and 3’ UTR sequences, yet it also depends on a correlation study, and its target is limited to endogenous mRNA sequences that are under evolutionary constraints (Bazzini et al., 2016; Narula et al., 2019; Wu et al., 2019). In contrast, PACE places each codon in a single artificial context to directly evaluate the codon effects on mRNA stability. As the findings of PACE in this study were largely consistent with those of CSC, our study provides additional, experimental validation of previously reported codon effects on mRNA stability in zebrafish (Figure 2B-E and S1D). Furthermore, we showed several lines of evidence indicating that codon effects stem from the decoding process in zebrafish. First, codons with destabilizing effects had fewer corresponding tRNA amounts than those with stabilizing effects (Figure 2D). Second, the occupancy of the ribosome with an A-site codon for a polar amino acid was negatively correlated with the codon’s effect on mRNA stability (Figure 3C) and the corresponding tRNA amount (Figure 3G). Third, experimental reduction of Asn tRNA availability conferred a destabilizing effect to the Asn AAC codon (Figure 4). Together, our study supports the view wherein a slow ribosome waiting for decoding by tRNAs triggers mRNA degradation in eukaryotes.\n",
      "31 One limitation of our analysis was that a correlation between the codon effect and the decoding process was not observed with codons for nonpolar and charged codons (Figure 3D and 3E). Although the exact reason is unknown, we noted that the ribosome occupancy and tRNA amounts were poorly correlated with these codons (Figure 3H and I). As either local or net hydrophobicity and the charge of the nascent peptides affects ribosome movement and functionality (Chadani et al., 2017, 2016; Charneski and Hurst, 2013; Dao Duc and Song, 2018; Lu and Deutsch, 2008; Weinberg et al., 2016), one possibility is that ribosome occupancy on these codon tags predominantly reflects the effect of nascent peptides; therefore, a precise measurement of the ribosome waiting for decoding is hampered. Future studies are necessary to confirm whether the decoding process of codons for nonpolar and charged amino acids also serves as a determinant of mRNA stability in zebrafish.\n",
      "32 Recent studies in yeast have provided mechanistic insights into codon-mediated decay and NGD, establishing them as distinct cotranslational mRNA decay pathways (Buschauer et al., 2020; D’Orazio et al., 2019; Ikeuchi et al., 2019; Webster et al., 2018). However, their definition is less clear in other organisms due to the lack of a connection between cis-elements and trans-acting factors. We showed that ribosome stall formed by hCMV uORF2 induced mRNA decay in a deadenylation-independent and Znf598-dependent manner, thus providing the first example of NGD in zebrafish embryos (Figure 5A-I). Irrespective of that, we found that the codon effects were largely maintained even in the absence of Znf598 (Figure 5J). Hence, cellular mechanisms clearly distinguish ribosome slowdown during the decoding process from ribosome stall by a problematic RNA or peptide and promote specific downstream mRNA decay pathways in zebrafish.\n",
      "33 How does the cell distinguish ribosomes for codon-mediated decay and those for NGD? Although earlier yeast studies provided the insights, it is unclear whether the underlying molecular mechanisms for recognizing slow ribosomes and stalled ribosomes are evolutionally conserved in higher eukaryotes. Whereas Not5 is not conserved in vertebrates, Not3 shares similarity with Not5 in the N-terminal domain required for E-site binding. It is therefore possible that the yeast Not3 ortholog CNOT3 anchors the CCR4-NOT complex to the slow ribosome via the E site in vertebrates (Buschauer et al., 2020; Collart and Panasenko, 2017). The ubiquitination site(s) on the ribosome and downstream NGD effectors are also conserved in vertebrates (D’Orazio et al., 2019; Garzia et al., 2017; Ikeuchi et al., 2019; Juszkiewicz et al., 2020, 2018; Juszkiewicz and Hegde, 2017; Matsuo et al., 2017; Sundaramoorthy et al., 2017; Weber et al., 2020). Future studies will reveal the conservation and biological function of codon-mediated decay and NGD. As exemplified by our analysis with the znf598 mutant, PACE will be a valuable tool to compare codon effects in multiple mutant backgrounds in such studies.\n",
      "34 Although the ribosome slows down locally on a destabilizing codon, our tandem ORF assay showed that such slowdown had a limited impact on the overall translation elongation rate (Figure 6). This assay was sensitive enough to detect the inhibitory effect of RQC-prone consecutive AAA codons. Therefore, we propose that ribosome slowdown during codon-mediated decay does not reach a nonproductive stall and maintains a productive translation cycle. This model is reasonable given that codon-mediated decay regulates the stability of thousands of endogenous protein-coding mRNAs for essential metabolic genes and maternal genes (Bazzini et al., 2016; Forrest et al., 2020; Mishima and Tomari, 2016; Presnyak et al., 2015). One remaining question in codon-mediated decay is how such transient slowness of the ribosome leads to robust deadenylation and mRNA degradation. In the case of RQC/NGD, a prolonged and robust ribosome stall causes a collision with the following ribosome, which becomes a substrate for Hel2/Znf598 (Ikeuchi et al., 2019; Juszkiewicz et al., 2018; Matsuo et al., 2020). While the collision itself is estimated to occur when the elongation rate decreases several-fold (∼6 codons/second to ∼1 codon/second) (Juszkiewicz et al., 2018), live imaging of RQC revealed that a much longer stall is required for RQC to occur (∼0.12 codons/second) (Goldman et al., 2021). Our study indicated that ribosome slowdown required for triggering codon-mediated decay did not reach such a prolonged ribosome stall (Figure 6). Among the two deadenylases in the Ccr4-Not complex, Caf1 is involved in deadenylation related to codon optimality in yeast (Webster et al., 2018). As Caf1 is a processive deadenylase, only a transient interaction between the Ccr4-Not complex and a slow ribosome, possibly mediated by E-site recognition, may be sufficient to initiate continuous deadenylation (Viswanathan et al., 2004). Alternatively, ubiquitination of eS7 by Not4 acts as an additional scaffold on the ribosome to stably recruit the Ccr4-Not complex after transient slowdown (Buschauer et al., 2020). Kinetical analysis of codon-mediated decay with a defined reporter mRNA is required to understand its detailed mechanism of action.\n",
      "35 The zebrafish AB strain was raised and maintained at 28.5°C under standard laboratory conditions according to the Animal Experiment Protocol (2018-46) at Kyoto Sangyo University. Fertilized eggs were obtained by natural breeding. Embryos were developed in system water at 28.5°C. Bright field images were acquired with a SteREO Lumar V12 microscope and an AxioCam MRc camera with ZEN software (Zeiss, Jena, Germany) using spline mode.\n",
      "36 To synthesize mRNA for microinjection, RNA was transcribed from a linearized plasmid DNA template using the SP6-Scribe Standard RNA IVT Kit (CELLSCRIPT) and purified with the RNeasy Mini Kit (QIAGEN). The m7G cap structure was added to the purified RNA using the ScriptCap m7G Capping System (CELLSCRIPT). Capped mRNA was purified with the RNeasy Mini Kit. For sfGFP reporter mRNAs, the poly(A) tail was added using the A-Plus Poly(A) Polymerase Tailing Kit (CELLSCRIPT), and the mRNA was then purified with the RNeasy Mini Kit. Purified mRNAs were diluted with water to the following concentrations: sfGFP reporter mRNAs and PACE reporter library 50 ng/μl; MT-Znf598 mRNAs 100 ng/μl; AnsB 50 ng/μl; tandem ORF reporter mRNAs 100 ng/μl. GFP MO was injected as described previously (Mishima and Tomari, 2016). Microinjection was performed using an IM300 Microinjector (NARISHIGE). Approximately 1,000 pl of solution was injected per embryo within 15 min after fertilization.\n",
      "37 sgRNA targeting exon 1 of zebrafish znf598 (ENSDARG00000014945) was designed with CRISPRscan (Moreno-Mateos et al., 2015). The DNA template for sgRNA synthesis was prepared by PCR using gene-specific primer and sgRNA tail primer. sgRNAs were transcribed in vitro using T7 RNA Polymerase (TaKaRa) and purified with ProbeQuant G-50 microcolumns (Cytiva). m7G-capped nCas9 mRNA was prepared with the SP6-Scribe Standard RNA IVT Kit and ScriptCap m7G Capping System (CELLSCRIPT) using pCS2+nCas9n as a template (Jao et al., 2013). sgRNA and nCas9 mRNA were mixed at 50 ng/μl each and injected into fertilized eggs. Embryos were raised to adulthood, and the fishes were crossed with AB fish. F1 embryos were screened by PCR analysis using Multina (SHIMADZU). DNA sequences were determined by cloning PCR fragments and Sanger sequencing. We identified a strain with an 11 bp deletion in exon 1, as shown in Figure 5F. Siblings of F1 fishes were raised to adulthood, and the fish harboring a heterozygous mutation were identified by fin-clipping and genotyping PCR. The F1 fish were crossed with AB to dilute any nonspecific mutations, and the F2 fish were used to obtain homozygous mutants. Primers are listed in Table S1\n",
      "38 To construct reporter plasmids, pCS2+ (Rupp et al., 1994) was modified by PCR to generate pCS2+neo using primers y683fw, y683rv, y684fw and y684rv. The suv39h1a 3’ UTR was excised from the previously described EGFP reporter plasmid (Mishima and Tomari, 2016) and cloned into pCS2+neo by XhoI and XbaI. The amino acid sequence of the sfGFP ORF was converted to nucleotide sequences comprising the most frequent synonymous codons in E. coli using EMBOSS Backtranseq (http://www.ebi.ac.uk/Tools/st/emboss_backtranseq/). The first 25 nucleotides of the sfGFP ORF were identical to the EGFP ORF used in a previous study to maintain the target site for GFP Morpholino oligo (Mishima and Tomari, 2016). EcoRI and EcoRV sites were inserted before the stop codon. The sfGFP DNA fragment was synthesized by GeneArt Strings DNA Fragments service (Life Technologies) and cloned into the pCS2+neo-suv39h1a 3’ UTR by NcoI and XhoI. DNA oligonucleotides encoding hCMV uORF2 were annealed and cloned into the pCS2+neo-sfGFP-suv39h1a 3’ UTR by EcoRI and EcoRV.\n",
      "39 To construct the tandem ORF reporter plasmid, the EGFP ORF was amplified by PCR using primers y670 and y671 and then amplified by y671 and y674. DsRedEx was amplified by PCR using y672 and y673 and then amplified by y675 and y673. The resulting EGFP fragment contained EGFP-P2A-Flag and the DsRedEx fragment contained P2A-HA-DsRedEx. Finally, the two fragments were joined by PCR using y670 and y673, generating EGFP-P2A-Flag-P2A-HA-DsRedEx. The final product was cloned into pCS2+MT containing the suv39 h1a 3’UTR using NcoI and XhoI sites. To insert the Lys AAG ×8 or AAA ×8 sequence, primers y713fw and y713rv or y714fw and y714rv were annealed and cloned downstream of Flag by EcoRV and EcoRI. Asn codontags were also cloned downstream of Flag by EcoRV and EcoRI. Optimal and nonoptimal EGFP described previously (Mishima and Tomari, 2016) were amplified by PCR using y707 and y708 and cloned downstream of Flag by EcoRV and EcoRI.\n",
      "40 Full-length and ΔRING Znf598 ORF fragments were amplified by RT-PCR using primers y624 and y775 and cloned into pCS2+MT by XhoI and XbaI. AnsB was amplified by PCR from the genomic DNA of E.coli DH5α using primers y603 and cloned into pCS2+SBP-HA (Mishima and Tomari, 2016) by EcoRI and XhoI. Primers are listed in Table S1.\n",
      "41 Oligonucleotides encoding codon tag sequences for all codons were synthesized as 64 pairs of single-strand DNA fragments. Complementary oligonucleotides were annealed and cloned into the pCS2+neo-sfGFP-suv39h1a 3’ UTR by EcoRI and EcoRV. The sequence of the cloned plasmids was confirmed by Sanger sequencing. Three codon tags, for UCU, UCA, and UAG, could not be cloned for unknown reasons. The cloned plasmids were mixed at an equimolar concentration and linearized by NotI. Capped and polyadenylated mRNAs were prepared using the mixed templates and purified as described above. The homogenous length of the PACE library mRNAs was confirmed by electrophoresis. The codon tag sequences are listed in Table S2.\n",
      "42 Total RNA was extracted from 40–50 embryos injected with the PACE library mRNAs using TRI Reagent (Molecular Research Center) and purified using the RNeasy Mini Kit (QIAGEN). After RNA integrity was confirmed using an Agilent RNA 6000 Nano Chip (Agilent Technologies, USA), total RNA samples were treated with Ribo-Zero (Illumina) and used for RNA-seq library preparation using an Illumina TruSeq Stranded Total RNA Library Prep Kit (Illumina, USA) according to the manufacturer’s instructions with multiplexing. The pooled libraries were sequenced on HiSeq4000 or NextSeq500 (Illumina) sequencing platforms with single-end sequencing of 65 (HiSeq4000) or 76 bp (NextSeq500) lengths.\n",
      "43 Ribosome footprint profiling using zebrafish embryos was performed as described previously (Han et al., 2020; Mito et al., 2020). Briefly, 50–60 zebrafish embryos injected with PACE reporter mRNAs were snap-frozen at the sphere stage and lysed with lysis buffer containing cycloheximide at 100 μg/ml (Sigma-Aldrich). An S-400 HR gel filtration spin column (GE Healthcare) was used to isolate the ribosomes. RNA fragments ranging from 26-34 nt were selected for monosome profiling. PCR-amplified DNA libraries were gel-excised and sequenced with the HiSeq4000 platform (Illumina).\n",
      "44 RNA-seq reads for PACE analysis were analyzed using Galaxy on public servers (Goecks et al., 2010). Single-ended reads matched to the codon tag sequences were counted by Salmon quant (Patro et al., 2017) with default parameters using the sequences listed in Table S2 as references. The number of reads on each codon tag was normalized to the total reads mapped to the zebrafish genome (GRCz10/danRer10) by TopHat v2.1.1. To calculate the stability of the codon tag reporter mRNAs, the normalized read numbers at 6 hpf were divided by the normalized read numbers at 2 hpf. The stability of each codon tag reporter mRNA was further normalized by the value of a UGA codon tag reporter mRNA. The results from two biological replicates were averaged and used for the analysis.\n",
      "45 Sequences from the ribosome profiling data were aligned against ncRNA sequences (GRCz11/danRer11.noncoding RNA genes data set) and mitochondrial gene sequences using Bowtie2; sequences that aligned were removed. The remaining sequences were aligned to the codon tag reporter sequences or the zebrafish genome (GRCz11/danRer11) using TopHat. Specific A-site assignment was determined by sequences aligned to the zebrafish genome, calibrated from footprints mapped to the beginning of the CDSes. The raw footprint read count at the test codon or spacer codon (as described in Figure 3A) was counted for further calculations.\n",
      "46 qRT-PCR analysis of mRNAs was performed as described previously (Mishima and Tomari, 2016). Briefly, total RNA was prepared using TRI Reagent (Molecular Research Center) and cDNA was synthesized using the PrimeScript RT reagent kit with gDNA eraser (TaKaRa). A random hexamer was used for cDNA synthesis. qRT-PCR was performed using SYBR Premix Ex TaqII (Tli RNaseH Plus) and the Thermal Cycler Dice Real Time System (TaKaRa). The results were analyzed based on standard curves from a serial dilution of cDNA. The data were normalized using 18S rRNA as a reference, and the mRNA level at 6 hpf was normalized to the data at 2 hpf. The primer sequences are listed in Table S1.\n",
      "47 To measure total tRNA and the charged tRNAs, we modified the OXOPAP assay (Gaston et al., 2008) and combined it with qRT-PCR analysis. Total RNA was purified with TRI reagent and dissolved in storage buffer (50 mM NaOAc pH 5.2 and 1 mM EDTA). For charged tRNA, 1 μg of the total RNA was incubated in oxidization buffer (50 mM NaIO4 and 100 mM NaOAc pH 5.2) supplemented with RNasin (Promega) for 30 min at 25°C to oxidize the nonacylated tRNA 3’ end. For the total tRNA, 1 μg of the total RNA was incubated in control buffer (50 mM NaCl and 100 mM NaOAc pH 5.2) supplemented with RNasin for 30 min at 25°C. Reactions were quenched with 100 mM glucose, incubated for 5 min at 25°C, and purified with G-25 columns (Cytiva). RNA samples were deacylated by adding 3.4 μl of 1 M Tris-HCl pH 9.0 and incubated for 30 min at 37°C. RNA samples were purified by ethanol precipitation and dissolved in 6.5 μl of RNase-free water. Four microliters of the RNA sample was denatured at 65°C for 5 min, and polyadenylated by the Poly(A) Polymerase Tailing Kit (CELLSCRIPT) in a 10 μl reaction in the presence of 2 U ePAP at 37°C for 60 min. Two-point five microliters of the polyadenylated sample was used for cDNA synthesis by the PrimeScript RT reagent kit with gDNA eraser (TaKaRa), with the y625 primer in Table S1. Reverse transcription was performed at 50°C for 30 min.\n",
      "48 qRT-PCR was performed as described above using the tRNA-specific forward primers shown in Table S1. Specific amplification was confirmed by dissociation curve analysis followed by gel electrophoresis and sequencing. The charged tRNA level relative to the total tRNA level was calculated by the ΔΔCt method (ΔCt of total tRNA sample – ΔCt of charged tRNA sample). The charged tRNA levels of the AnsB-overexpression experiment were normalized to those of the mock injection experiment.\n",
      "49 The PAT assay was performed as described previously (Mishima and Tomari, 2016). Briefly, 150 ng of total RNA was incubated with 75 U of yeast poly(A) polymerase (PAP) (Affymetrix) in the presence of GTP/ITP mix at 37°C for 60 min. cDNA was synthesized at 44°C for 15 min using the PrimeScript RT reagent kit with gDNA eraser (TaKaRa) and a y300 PAT universal C10 primer. PAT-PCR was performed using a 3’ UTR-specific forward primer and a y300 PAT universal C10 primer with GoTaq Green Master Mix (Promega). The primer sequences are provided in Table S1. The PCR products were separated by 6% PAGE in 0.5×TBE. Gels were stained with GelRed (Biotium) and the signals were detected using LAS4000 (GE Healthcare) or Amersham Imager 680 (GE Healthcare). Signals were quantified using the “Plot profiles” function of the ImageJ software (http://imagej.nih.gov/ij/).\n",
      "50 Proteins were detected using anti-Myc (MBL My3 mouse monoclonal, 1:4,000), anti-HA (MBL TANA2 mouse monoclonal 1:4,000), anti-FLAG (MBL FLA-1 mouse monoclonal 1:4,000), anti-GFP (MBL No. 598 rabbit polyclonal, 1:3000), and anti-α-tubulin-pAb HRP-DirecT (MBL PM054-7 rabbit polyclonal 1: 10,000) antibodies. The signals were developed using Luminata Forte (Millipore) and detected using LAS4000 (GE Healthcare) or Amersham Imager 680 (GE Healthcare). Signals were quantified using the “Measure” function of the ImageJ software (http://imagej.nih.gov/ij/).\n",
      "51 The accession number for the sequence data reported in this paper is GEO: GSE173179 and GSE173604.\n"
     ]
    }
   ],
   "source": [
    "for i, p in enumerate(preprint_chunks):\n",
    "    print(i, p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ae21c58-1780-417d-b6fe-a63f20e958cc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
